kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ValidateDomainMetadataIterator.java

```diff
@@ -0,0 +1,94 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.util;
+
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This iterator wraps around a data action {@link CloseableIterator} and performs additional checks
+ * for {@link DomainMetadata} actions.
+ *
+ * <ul>
+ *   <li>Ensures domain metadata actions are only present when supported by the table protocol.
+ *   <li>Validates that there are no duplicate domain metadata actions for the same domain in the
+ *       provided actions.
+ *   <li>Throws an exception if the above requirements are not met.
+ * </ul>
+ */
+public class ValidateDomainMetadataIterator implements CloseableIterator<Row> {
+  private final CloseableIterator<Row> dataActionsIter;
+  private final Map<String, DomainMetadata> domainMetadataMap = new HashMap<>();
+  private final int domainMetadataOrdinal;
+  private final boolean domainMetadataSupported;
+  private final boolean allowDuplicate;
+
+  public ValidateDomainMetadataIterator(
+      Protocol protocol,
+      CloseableIterator<Row> dataActionsIter,
+      StructType schema,
+      boolean allowDuplicate) {
+    this.domainMetadataSupported = DomainMetadataUtils.isDomainMetadataSupported(protocol);
+    this.dataActionsIter = dataActionsIter;
+    this.allowDuplicate = allowDuplicate;
+    this.domainMetadataOrdinal = schema.indexOf("domainMetadata");
+    if (domainMetadataOrdinal == -1) {
+      throw new IllegalArgumentException("Schema does not contain 'domainMetadata' field");
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    return dataActionsIter.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    Row action = dataActionsIter.next();
+    if (!action.isNullAt(domainMetadataOrdinal)) {
+      // This is a DomainMetadata action
+      if (!domainMetadataSupported) {
+        throw DeltaErrors.domainMetadataUnsupported();
+      }
+
+      // Extract DomainMetadata action
+      DomainMetadata domainMetadata =
+          DomainMetadata.fromRow(action.getStruct(domainMetadataOrdinal));
+
+      // Check if domain already seen
+      String domain = domainMetadata.getDomain();
+      if (!allowDuplicate && domainMetadataMap.containsKey(domain)) {
+        throw DeltaErrors.duplicateDomainMetadataAction(
+            domainMetadataMap.get(domain).toString(), domainMetadata.toString());
+      }
+      domainMetadataMap.put(domain, domainMetadata);
+    }
+    return action;
```

Comment:
what if action is null at domainMetadataOrdinal? you just return it? that doesn't seem right to me ...

Reply:
Hmmm... It seems to be the desired behavior to me.

This iterator's purpose is to specifically validate `DomainMetadata` actions while allowing all other action types to pass through unchanged. When the action is null at `domainMetadataOrdinal`, it indicates this is a non-`DomainMetadata` action, so I think we should just return it.

Reply:
I understand now, thanks!